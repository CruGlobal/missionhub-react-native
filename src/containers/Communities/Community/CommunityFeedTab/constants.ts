export const COMMUNITY_FEED = 'nav/COMMUNITY_FEED';
