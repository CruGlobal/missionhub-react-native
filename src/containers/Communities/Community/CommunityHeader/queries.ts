import gql from 'graphql-tag';

export const COMMUNITY_HEADER_QUERY = gql`
  query CommunityHeader($id: ID!, $myId: ID!) {
    community(id: $id) {
      id
      name
      communityPhotoUrl
      report(period: "P99Y") {
        memberCount
      }
      people(personIds: [$myId]) {
        edges {
          communityPermission {
            permission
          }
        }
      }
    }
  }
`;

export const COMMUNITY_HEADER_GLOBAL_QUERY = gql`
  query CommunityHeaderGlobal {
    globalCommunity {
      usersReport {
        usersCount
      }
    }
  }
`;
