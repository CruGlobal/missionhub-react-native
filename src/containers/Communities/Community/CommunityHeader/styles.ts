import { StyleSheet } from 'react-native';

import theme from '../../../../theme';

export default StyleSheet.create({
  container: {
    shadowColor: theme.black,
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.3,
    shadowRadius: 2,
    backgroundColor: theme.grey3,
  },
  image: { height: 210 },
  imageOverlay: {
    ...StyleSheet.absoluteFillObject,
    backgroundColor: theme.black,
    opacity: 0.15,
  },
  communityName: {
    fontSize: 24,
    fontWeight: '300',
    lineHeight: 30,
    color: theme.white,
  },
  communityMembersButton: {
    paddingHorizontal: 12,
    paddingVertical: 3,
    borderRadius: 30,
    borderColor: theme.white,
    borderWidth: 1,
    marginBottom: 10,
  },
  communityMembersText: {
    fontWeight: 'normal',
    fontSize: 14,
    lineHeight: 20,
    color: theme.white,
  },
});
