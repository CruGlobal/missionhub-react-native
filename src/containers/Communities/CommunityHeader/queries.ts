import gql from 'graphql-tag';

export const COMMUNITY_HEADER_QUERY = gql`
  query CommunityHeader($id: ID!) {
    community(id: $id) {
      name
      communityPhotoUrl
      report(period: "P99Y") {
        memberCount
      }
    }
  }
`;
