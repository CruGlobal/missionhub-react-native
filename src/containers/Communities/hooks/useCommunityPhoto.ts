import { ImageSourcePropType } from 'react-native';

import { orgIsGlobal } from '../../../utils/common';
import GLOBAL_COMMUNITY_IMAGE from '../../../../assets/images/globalCommunityImage.png';

export const useCommunityPhoto = (
  communityId: string,
  communityPhotoUrl?: string | null,
): ImageSourcePropType => {
  if (communityPhotoUrl) {
    return { uri: communityPhotoUrl };
  } else if (orgIsGlobal({ id: communityId })) {
    return GLOBAL_COMMUNITY_IMAGE;
  } else {
    return {};
  }
};
