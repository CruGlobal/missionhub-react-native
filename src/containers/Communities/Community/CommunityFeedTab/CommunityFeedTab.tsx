import React, { useContext } from 'react';
import { useSelector } from 'react-redux';
import { useNavigationParam } from 'react-navigation-hooks';

import { CollapsibleViewContext } from '../../../../components/CollapsibleView/CollapsibleView';
import { CommunityFeed } from '../../../CommunityFeed';
import { useAnalytics } from '../../../../utils/hooks/useAnalytics';
import { getAnalyticsPermissionType } from '../../../../utils/analytics';
import { organizationSelector } from '../../../../selectors/organizations';
import { OrganizationsState } from '../../../../reducers/organizations';
import { AuthState } from '../../../../reducers/auth';
import { ANALYTICS_PERMISSION_TYPE } from '../../../../constants';
import { orgIsGlobal } from '../../../../utils/common';

interface CommunityFeedTabProps {
  collapsibleHeaderContext: CollapsibleViewContext;
}

export const CommunityFeedTab = ({
  collapsibleHeaderContext,
}: CommunityFeedTabProps) => {
  const communityId: string = useNavigationParam('communityId');
  const personId: string | undefined = useNavigationParam('personId');

  const organization = useSelector(
    ({ organizations }: { organizations: OrganizationsState }) =>
      organizationSelector({ organizations }, { orgId: communityId }),
  );

  const analyticsPermissionType = useSelector(({ auth }: { auth: AuthState }) =>
    getAnalyticsPermissionType(auth, organization),
  );

  useAnalytics('community feed', {
    screenContext: { [ANALYTICS_PERMISSION_TYPE]: analyticsPermissionType },
  });

  const { collapsibleScrollViewProps } = useContext(collapsibleHeaderContext);

  return (
    <CommunityFeed
      testID="CelebrateFeed"
      communityId={communityId}
      personId={personId}
      noHeader={!!personId}
      itemNamePressable={!orgIsGlobal(organization)}
      collapsibleScrollViewProps={collapsibleScrollViewProps}
    />
  );
};
