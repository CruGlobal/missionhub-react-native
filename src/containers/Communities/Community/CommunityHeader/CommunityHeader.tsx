import React from 'react';
import { View, ImageBackground } from 'react-native';
import { useTranslation } from 'react-i18next';
import { useQuery } from '@apollo/react-hooks';
import { useDispatch } from 'react-redux';
import { useNavigationParam } from 'react-navigation-hooks';

import {
  createCollapsibleViewContext,
  CollapsibleViewHeader,
} from '../../../../components/CollapsibleView/CollapsibleView';
import { Text, Flex, Button } from '../../../../components/common';
import { HeaderTabBar } from '../../../../components/HeaderTabBar/HeaderTabBar';
import BackButton from '../../../BackButton';
import theme from '../../../../theme';
import Header from '../../../../components/Header';
import { ErrorNotice } from '../../../../components/ErrorNotice/ErrorNotice';
import { navigatePush } from '../../../../actions/navigation';
import { COMMUNITY_PROFILE } from '../CommunityProfile/CommunityProfile';
import { COMMUNITY_MEMBERS } from '../CommunityMembers/CommunityMembers';
import InfoIcon from '../../../../../assets/images/infoIcon.svg';
import EditIcon from '../../../../../assets/images/editIcon.svg';
import { useMyId } from '../../../../utils/hooks/useIsMe';
import { useCommunityPhoto } from '../../hooks/useCommunityPhoto';
import { canEditCommunity, orgIsGlobal } from '../../../../utils/common';
import { communityTabs } from '../constants';

import styles from './styles';
import {
  COMMUNITY_HEADER_QUERY,
  COMMUNITY_HEADER_GLOBAL_QUERY,
} from './queries';
import {
  CommunityHeader as CommunityHeaderQuery,
  CommunityHeaderVariables,
} from './__generated__/CommunityHeader';
import { CommunityHeaderGlobal } from './__generated__/CommunityHeaderGlobal';

export const CommunitiesCollapsibleHeaderContext = createCollapsibleViewContext();

export const CommunityHeader = () => {
  const { t } = useTranslation('communityHeader');
  const dispatch = useDispatch();
  const communityId: string = useNavigationParam('communityId');

  const myId = useMyId();
  const isGlobalCommunity = orgIsGlobal({ id: communityId });

  const { data, error, refetch } = useQuery<
    CommunityHeaderQuery,
    CommunityHeaderVariables
  >(COMMUNITY_HEADER_QUERY, {
    variables: { id: communityId, myId },
    skip: isGlobalCommunity,
  });
  const {
    data: globalData,
    error: globalError,
    refetch: globalRefetch,
  } = useQuery<CommunityHeaderGlobal>(COMMUNITY_HEADER_GLOBAL_QUERY, {
    skip: !isGlobalCommunity,
  });

  const communityPhotoSource = useCommunityPhoto(
    communityId,
    data?.community.communityPhotoUrl,
    data?.community.userCreated,
  );

  return (
    <CollapsibleViewHeader
      context={CommunitiesCollapsibleHeaderContext}
      headerHeight={260}
    >
      <View style={styles.container}>
        <ImageBackground source={communityPhotoSource} style={styles.image}>
          <View style={styles.imageOverlay} />
          <Header
            left={<BackButton />}
            right={
              isGlobalCommunity ? null : (
                <Button
                  testID="communityProfileButton"
                  onPress={() =>
                    dispatch(navigatePush(COMMUNITY_PROFILE, { communityId }))
                  }
                >
                  {data?.community.userCreated &&
                  canEditCommunity(
                    data?.community.people.edges[0].communityPermission
                      .permission,
                    data?.community.userCreated,
                  ) ? (
                    <EditIcon color={theme.white} />
                  ) : (
                    <InfoIcon color={theme.white} />
                  )}
                </Button>
              )
            }
          />
          <Flex
            value={1}
            justify="around"
            style={{
              paddingLeft: 25,
              paddingRight: 100,
            }}
          >
            <Text style={styles.communityName} numberOfLines={2}>
              {isGlobalCommunity
                ? t('globalCommunity')
                : data?.community.name ?? '-'}
            </Text>
            <Flex direction="row" align="start">
              <Button
                testID="communityMembersButton"
                pill={true}
                type="transparent"
                onPress={() =>
                  !isGlobalCommunity &&
                  dispatch(navigatePush(COMMUNITY_MEMBERS, { communityId }))
                }
                style={styles.communityMembersButton}
                buttonTextStyle={styles.communityMembersText}
                text={t('memberCount', {
                  count:
                    (isGlobalCommunity
                      ? globalData?.globalCommunity.usersReport.usersCount
                      : data?.community.report.memberCount) ?? 0,
                })}
              />
            </Flex>
          </Flex>
        </ImageBackground>
        <ErrorNotice
          message={t('errorLoadingCommunityDetails')}
          error={error}
          refetch={refetch}
        />
        <ErrorNotice
          message={t('errorLoadingCommunityDetails')}
          error={globalError}
          refetch={globalRefetch}
        />
        <HeaderTabBar tabs={communityTabs} />
      </View>
    </CollapsibleViewHeader>
  );
};
