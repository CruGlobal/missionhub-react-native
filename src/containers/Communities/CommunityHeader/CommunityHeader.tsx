import React from 'react';
import { View, ImageBackground } from 'react-native';
import { useTranslation } from 'react-i18next';
import { useQuery } from '@apollo/react-hooks';
import { useDispatch } from 'react-redux';

import {
  createCollapsibleViewContext,
  CollapsibleViewHeader,
} from '../../../components/CollapsibleView/CollapsibleView';
import { Text, Flex, Button } from '../../../components/common';
import { HeaderTabBar } from '../../../components/HeaderTabBar/HeaderTabBar';
import {
  GROUP_CELEBRATE,
  GROUP_CHALLENGES,
  GROUP_IMPACT,
} from '../../Groups/GroupScreen';
import BackButton from '../../BackButton';
import theme from '../../../theme';
import Header from '../../../components/Header';
import { ErrorNotice } from '../../../components/ErrorNotice/ErrorNotice';
import { navigatePush } from '../../../actions/navigation';
import { COMMUNITY_PROFILE } from '../CommunityProfile/CommunityProfile';
import InfoIcon from '../../../../assets/images/infoIcon.svg';
import EditIcon from '../../../../assets/images/editIcon.svg';
import { useMyId } from '../../../utils/hooks/useIsMe';
import { useCommunityPhoto } from '../hooks/useCommunityPhoto';
import { canEditCommunity } from '../../../utils/common';

import styles from './styles';
import { COMMUNITY_HEADER_QUERY } from './queries';
import {
  CommunityHeader as CommunityHeaderQuery,
  CommunityHeaderVariables,
} from './__generated__/CommunityHeader';

export const CommunitiesCollapsibleHeaderContext = createCollapsibleViewContext();

interface CommunityHeaderProps {
  communityId: string;
}

export const CommunityHeader = ({ communityId }: CommunityHeaderProps) => {
  const { t } = useTranslation('communityHeader');
  const dispatch = useDispatch();
  const myId = useMyId();

  const { data, error, refetch } = useQuery<
    CommunityHeaderQuery,
    CommunityHeaderVariables
  >(COMMUNITY_HEADER_QUERY, {
    variables: { id: communityId, myId },
  });

  const communityPhotoSource = useCommunityPhoto(
    communityId,
    data?.community.communityPhotoUrl,
    data?.community.userCreated,
  );

  return (
    <CollapsibleViewHeader
      context={CommunitiesCollapsibleHeaderContext}
      headerHeight={260}
    >
      <View style={styles.container}>
        <ImageBackground source={communityPhotoSource} style={styles.image}>
          <View style={styles.imageOverlay} />
          <Header
            left={<BackButton />}
            right={
              <Button
                onPress={() =>
                  dispatch(navigatePush(COMMUNITY_PROFILE, { communityId }))
                }
              >
                {canEditCommunity(
                  data?.community.people.edges[0].communityPermission
                    .permission,
                  data?.community.userCreated,
                ) ? (
                  <EditIcon color={theme.white} />
                ) : (
                  <InfoIcon color={theme.white} />
                )}
              </Button>
            }
          />
          <Flex
            value={1}
            justify="around"
            style={{
              paddingLeft: 25,
              paddingRight: 100,
            }}
          >
            <Text style={styles.communityName} numberOfLines={2}>
              {data?.community.name ?? '-'}
            </Text>
            <Flex direction="row" align="start">
              <Button
                pill={true}
                type="transparent"
                onPress={() => {
                  // TODO: navigate to new members list screen
                }}
                style={styles.communityMembersButton}
                buttonTextStyle={styles.communityMembersText}
                text={t('memberCount', {
                  count: data?.community.report.memberCount ?? 0,
                })}
              />
            </Flex>
          </Flex>
        </ImageBackground>
        <ErrorNotice
          message={t('errorLoadingCommunityDetails')}
          error={error}
          refetch={refetch}
        />
        <HeaderTabBar
          tabs={[
            { name: t('feed'), navigationAction: GROUP_CELEBRATE },
            {
              name: t('challenges'),
              navigationAction: GROUP_CHALLENGES,
            },
            { name: t('impact'), navigationAction: GROUP_IMPACT },
          ]}
        />
      </View>
    </CollapsibleViewHeader>
  );
};
