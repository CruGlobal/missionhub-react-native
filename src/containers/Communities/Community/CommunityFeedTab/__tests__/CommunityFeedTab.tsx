import React from 'react';

import { renderWithContext } from '../../../../../../testUtils';
import { CommunityFeedTab } from '../CommunityFeedTab';
import { CommunitiesCollapsibleHeaderContext } from '../../CommunityHeader/CommunityHeader';
import { useAnalytics } from '../../../../../utils/hooks/useAnalytics';
import { PermissionEnum } from '../../../../../../__generated__/globalTypes';
import { getAnalyticsPermissionType } from '../../../../../utils/analytics';

jest.mock('../../../../CommunityFeed', () => ({
  CommunityFeed: 'CommunityFeed',
}));
jest.mock('../../../../../utils/analytics');
jest.mock('../../../../../utils/hooks/useAnalytics');

const communityId = '1';
const personId = '2';
const permissionType = PermissionEnum.owner;

const initialState = {
  organizations: { all: [{ id: communityId }] },
};

(getAnalyticsPermissionType as jest.Mock).mockReturnValue(permissionType);

it('should render correctly', () => {
  renderWithContext(
    <CommunityFeedTab
      collapsibleHeaderContext={CommunitiesCollapsibleHeaderContext}
    />,
    { initialState, navParams: { communityId, personId } },
  ).snapshot();

  expect(useAnalytics).toHaveBeenCalledWith('community feed', {
    permissionType: { communityId },
  });
});
