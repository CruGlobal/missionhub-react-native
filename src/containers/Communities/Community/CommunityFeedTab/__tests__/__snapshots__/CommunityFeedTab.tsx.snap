// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<CommunityFeed
  communityId="1"
  itemNamePressable={true}
  noHeader={true}
  personId="2"
  testID="CelebrateFeed"
/>
`;
